scipy/stats/_multivariate.py

```diff
@@ -2642,3 +2644,104 @@ def entropy(self):
     method_frozen.__doc__ = doccer.docformat(
         method.__doc__, wishart_docdict_noparams)
     method.__doc__ = doccer.docformat(method.__doc__, wishart_docdict_params)
+
+class rotation_gen(multi_rv_generic):
+    r"""
+    A matrix-valued SO(N) random variable.
+
+    The `dim` keyword specifies the dimension N.
+
+    Methods
+    -------
+    ``rvs(dim=None, size=1, random_state=None)``
+        Draw random samples from SO(N).
+
+    Parameters
+    ----------
+    dim : scalar
+        Dimension of matrices
+
+    TODO: frozen object not implemented yet.
+
+    Notes
+    -----
+    Return a random rotation matrix, drawn from the Haar distribution
+    (the only uniform distribution on SO(n)).
+
+    Examples
+    --------
+    TODO: Add examples
+
+    """
+
+    def __init__(self, seed=None):
+        super(rotation_gen, self).__init__(seed)
+        self.__doc__ = doccer.docformat(self.__doc__)
+
+    # TODO: Define __call__ for frozen dist'n
+
+    def _process_parameters(self, dim):
+        """
+        Dimension N must be specified; it cannot be inferred.
+        """
+
+        if dim is None or not np.isscalar(dim) or dim <= 1:
+          raise ValueError("Dimension of rotation must be specified, and must be a scalar greater than 1.")
+
+        return dim
+
+    def rvs(self, dim, size=1, random_state=None):
+        """
+        Draw random samples from SO(N).
+
+        Parameters
+        ----------
+        dim : integer
+            Dimension of rotation space (N).
+        size : integer, optional
+            Number of samples to draw (default 1).
+
+        Returns
+        -------
+        rvs : ndarray or scalar
+            Random size N-dimensional matrices, dimension (size, dim, dim)
+
+        Notes
+        -----
+        Return a random rotation matrix, drawn from the Haar distribution
+        (the only uniform distribution on SO(n)).
+        The algorithm is described in the paper
+        Stewart, G.W., "The efficient generation of random orthogonal
+        matrices with an application to condition estimators", SIAM Journal
+        on Numerical Analysis, 17(3), pp. 403-409, 1980.
+        For more information see
+        http://en.wikipedia.org/wiki/Orthogonal_matrix#Randomization
+        """
+        size = int(size)
+        if size > 1:
+          return np.array([ self.rvs(dim, size=1, random_state=random_state) for i in range(size) ])
```

Comment:
Yes, I noticed that after submitting it. Easily enough fixed.

On 23 December 2015 at 18:35, Robert Kern <notifications@github.com> wrote:

> In scipy/stats/_multivariate.py
> <https://github.com/scipy/scipy/pull/5622#discussion_r48387247>:
>
> > +            Random size N-dimensional matrices, dimension (size, dim, dim)
> > +
> > +        Notes
> > +        -----
> > +        Return a random rotation matrix, drawn from the Haar distribution
> > +        (the only uniform distribution on SO(n)).
> > +        The algorithm is described in the paper
> > +        Stewart, G.W., "The efficient generation of random orthogonal
> > +        matrices with an application to condition estimators", SIAM Journal
> > +        on Numerical Analysis, 17(3), pp. 403-409, 1980.
> > +        For more information see
> > +        http://en.wikipedia.org/wiki/Orthogonal_matrix#Randomization
> > +        """
> > +        size = int(size)
> > +        if size > 1:
> > +          return np.array([ self.rvs(dim, size=1, random_state=random_state) for i in range(size) ])
>
> We use 4-space indents.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy/scipy/pull/5622/files#r48387247>.
>